rust/bridge/node/src/lib.rs

```diff
@@ -3,13 +3,289 @@
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
+use async_trait::async_trait;
+use libsignal_bridge::node;
+use libsignal_protocol::*;
+use neon::context::Context;
 use neon::prelude::*;
+use signal_neon_futures::*;
+use std::fmt;
+use std::marker::PhantomData;
+use std::panic::AssertUnwindSafe;
+use std::sync::Arc;
 
 pub mod logging;
 
+#[derive(Debug)]
+struct CallbackError {
+    message: String,
+}
+
+impl CallbackError {
+    fn new(message: String) -> CallbackError {
+        Self { message }
+    }
+}
+
+impl fmt::Display for CallbackError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "callback error {}", self.message)
+    }
+}
+
+impl std::error::Error for CallbackError {}
+
+fn js_error_to_rust(func: &'static str, err: String) -> SignalProtocolError {
+    SignalProtocolError::ApplicationCallbackError(func, Box::new(CallbackError::new(err)))
+}
+
+pub fn boxed_object<'a, T: 'static + Send>(
+    cx: &mut TaskContext<'a>,
+    value: T,
+) -> JsResult<'a, JsValue> {
+    Ok(cx.boxed(node::DefaultFinalize(value)).upcast())
+}
+
+struct NodeSenderKeyStore<'a> {
+    js_queue: EventQueue,
+    store_object: Arc<Root<JsObject>>,
+    phantom: PhantomData<&'a ()>,
+}
+
+impl<'a> NodeSenderKeyStore<'a> {
+    fn new(cx: &mut FunctionContext, store: Handle<JsObject>) -> Self {
+        Self {
+            js_queue: cx.queue(),
+            store_object: Arc::new(store.root(cx)),
+            phantom: PhantomData,
+        }
+    }
+
+    async fn do_get_sender_key(
+        &self,
+        name: SenderKeyName,
+    ) -> Result<Option<SenderKeyRecord>, String> {
+        eprintln!("here...");
+        let store_object_shared = self.store_object.clone();
+        JsFuture::get_promise(&self.js_queue, move |cx| {
+            let store_object = store_object_shared.to_inner(cx);
+            let name: Handle<JsValue> = boxed_object(cx, name.clone())?; // XXX clone
```

Comment:
I don't have a good answer for how to avoid *this* clone, since we're handing brand new objects out to JS. I think most of the time it's fine, though: the callbacks all use small-ish objects like addresses and identity keys, or they're doing something like "save session" where we can transfer the object we were working with locally.